### ISSUE TITLE
Download button for single file.

### ISSUE THREAD

**Issue by Mayccoll** [buunguyen/octotree#134]
hi,
Some time ago to create a script that creates a button to download a single file that looks like this.

![download](http://i.imgur.com/MGfYHL9l.jpg)

For me it is very useful because sometimes I just need a single file and not the entire repository

I planned to create an add-on for Firefox until I came across Octotree and thought maybe you could include it in this add-on and would be something like this.

![download-octotree](http://i.imgur.com/5uxymnRl.jpg)

Here is my code

```js
function insert_btn () {

    var pathname = $(location).attr('href');
    var clonedown = $( "#raw-url" ).clone();

    clonedown.attr({
        id:"down_btn",
        href:pathname,
        download:""
    }).text("Download").insertAfter( "#raw-url" );
}

insert_btn();

window.setInterval(function(){
    if ($('#down_btn').length) {

    } else {
        insert_btn();
    }
}, 5000);
```
tnks

**Comment by pallavagarwal07** [comment]
I think this would be a very helpful feature. I'll try to do this if I can.

**Comment by joedf** [comment]
+1

**Comment by pallavagarwal07** [comment]
This doesn't seem to be possible.
@Mayccoll Your code downloads the http webpage, not the file itself. The url used should be that of the raw file. That however causes a redirect to githubusercontent domain and prevents "download" attribute from working.

**Comment by pallavagarwal07** [comment]
If anybody has any suggestion, it'd be very helpful

**Comment by Mayccoll** [comment]
Hi,
here is the new code working, i change the host for a CDN, this way you can download the file.
I'm not sure if is the best approach but is working fine :)

```js
function insert_btn () {

    var pathname   = $(location).attr('href');

    var removeHost = pathname.replace("github.com", "cdn.rawgit.com");

    var removeBlob = removeHost.replace("blob/", "");

    var filename   = pathname.replace(/^.*[\\\/]/, '');

    var clonedown  = $( "#raw-url" ).clone();

    clonedown.attr({
        id      : "down_btn",
        href    : removeBlob,
        download: filename,
        target  : "_blank"
    }).text("Download").insertAfter( "#raw-url" );
}

insert_btn();

window.setInterval(function(){
    if ($('#down_btn').length) {

    } else {
        insert_btn();
    }
}, 4000);
```

**Comment by pallavagarwal07** [comment]
I have a method that works perfectly, but involves routing everything through my personal domain for some server side magic. 
Not sure if the admin would like that

**Comment by joedf** [comment]
There's no need for all that.
You can use [`blob`](https://developer.mozilla.org/en/docs/Web/API/Blob).

Here's a simple example
```js
var responseText = getDataFromURL_somehow();  // ok lets say this var contains the data
var blob = new Blob([responseText], {type: 'application/octet-binary'});
var url = URL.createObjectURL(blob);
```

Here's another example : [permalink](https://github.com/Ragnar-F/AutoHotkey_L-Docs/blob/ef000631a3935c69b1bda61bd48562c1ff46f728/docs/static/content.js#L180-L209)

---
As a side note, there's also [Downloadify](https://github.com/dcneiner/Downloadify)... Not sure if it still works though.

**Comment by joedf** [comment]
Looks like there's a really cool library here for this.
https://github.com/eligrey/FileSaver.js

**Comment by buunguyen** [comment]
@joedf :+1: 

@pallavagarwal07 the GitHub adapter already includes [a method](https://github.com/buunguyen/octotree/blob/master/src/adapter.github.js#L204) to get the blob of a file. You should be able to combine it with the examples by @joedf to build this feature.

**Comment by lite3** [comment]
like this https://github.com/lite3/github-file-download/blob/master/github-file-download/download-button.js#L59-L69

HTML5 `download` attr will download a file.

**Comment by buunguyen** [comment]
@pallavagarwal07 if you're still interested in working on this, there's an easy way to get the raw file: just replace `/blob/` with `/raw/`. Then create the link node with the `download` attribute as you were doing. 

For example:
https://github.com/buunguyen/octotree/blob/master/src%2Foctotree.js vs https://github.com/buunguyen/octotree/raw/master/src%2Foctotree.js

**Comment by joedf** [comment]
**Working single function:**
```javascript
function dnl_blob(u) {
	var req = new XMLHttpRequest();
	req.onreadystatechange = function() {
		if (req.readyState == 4 && req.status == 200) {
			var responseText = req.responseText;
			var blob = new Blob([responseText], {type: 'application/octet-binary'});
			var blobURL = URL.createObjectURL(blob);
			alert(blobURL); //display download url
			window.location = blobURL; //Start the download
		}
    }; 
	req.open("GET",u,true); //asynchronous
	req.send();
}
```
**!Note:** Must be on the same domain, otherwise it is considered as XSS.

**Comment by buunguyen** [comment]
IMO, creating a link obj and set href to url.replace(/\/blob\/, '/raw/') is far simpler.

**Comment by joedf** [comment]
or that. :wink: 
Agreed.

**Comment by buunguyen** [comment]
Done. Should be available in 1.7.0 to Chrome and Firefox soon.

**Comment by joedf** [comment]
Great! :+1:
